pkg/metrics/client/client.go

```diff
@@ -0,0 +1,102 @@
+
+	prometheusapi "github.com/prometheus/client_golang/api"
+	prometheusv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+
+	routeclient "github.com/openshift/client-go/route/clientset/versioned"
+	routev1client "github.com/openshift/client-go/route/clientset/versioned/typed/route/v1"
+)
+
+// DEPRECATED: This function is left only for test library compatibility.
+func DeprecatedPrometheusClient(kubeClient *kubernetes.Clientset, routeClient *routeclient.Clientset) (prometheusv1.API, error) {
+	return NewPrometheusClient(kubeClient.CoreV1(), kubeClient.CoreV1(), kubeClient.CoreV1(), routeClient.RouteV1())
+}
+
+// NewPrometheusClient returns Prometheus API or error
+// Note: with thanos-querier you must pass an entire Alert as a query. Partial queries return an error, so have to pass the entire alert.
+// Example query for an Alert:
+// `ALERTS{alertname="PodDisruptionBudgetAtLimit",alertstate="pending",namespace="pdbnamespace",poddisruptionbudget="pdbname",prometheus="openshift-monitoring/k8s",service="kube-state-metrics",severity="warning"}==1`
+// Example query:
+// `scheduler_scheduling_duration_seconds_sum`
+func NewPrometheusClient(serviceClient corev1client.ServicesGetter, secretsClient corev1client.SecretsGetter, configsClient corev1client.ConfigMapsGetter, routeClient routev1client.RoutesGetter) (prometheusv1.API, error) {
+	_, err := serviceClient.Services("openshift-monitoring").Get("prometheus-k8s", metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	route, err := routeClient.Routes("openshift-monitoring").Get("thanos-querier", metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	host := route.Status.Ingress[0].Host
+	var bearerToken string
+	secrets, err := secretsClient.Secrets("openshift-monitoring").List(metav1.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("could not list secrets in openshift-monitoring namespace")
+	}
+	for _, s := range secrets.Items {
+		if s.Type != corev1.SecretTypeServiceAccountToken ||
+			!strings.HasPrefix(s.Name, "prometheus-k8s") {
+			continue
+		}
+		bearerToken = string(s.Data[corev1.ServiceAccountTokenKey])
+		break
+	}
+	if len(bearerToken) == 0 {
+		return nil, fmt.Errorf("prometheus service account not found")
+	}
+
+	return createClient(configsClient, host, bearerToken)
+}
+
+func createClient(configClient corev1client.ConfigMapsGetter, host, bearerToken string) (prometheusv1.API, error) {
+	// retrieve router CA
+	routerCAConfigMap, err := configClient.ConfigMaps("openshift-config-managed").Get("router-ca", metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	bundlePEM := []byte(routerCAConfigMap.Data["ca-bundle.crt"])
+
+	// make a client connection configured with the provided bundle.
+	roots := x509.NewCertPool()
+	roots.AppendCertsFromPEM(bundlePEM)
+
+	// prometheus API client, configured for route host and bearer token auth
+	client, err := prometheusapi.NewClient(prometheusapi.Config{
+		Address: "https://" + host,
```

Comment:
@deads2k @russellb will this break in ipv6 ?

Reply:
If `host` is ever an IPv6 address, then yes, it will break.

The safe way to handle this would be something like

`"https://" + net.JoinHostPort(host, "443")`

Reply:
^^^ will ensure that an IPv6 address gets wrapped in square brackets.

Reply:
@russellb this will be a host name (route host name)

Reply:
OK, then no problems.